### ISSUE TITLE
Feature request: Package provider for FreeBSD

### ISSUE THREAD

**Issue by kraileth** [comtrya/comtrya#70]
A little background for any reader who might not be familiar with Unix-likes besides Linux:

Modern FreeBSD and DragonFly BSD use a package manager called _pkg_ which has been available as an alternative from FreeBSD 8.4 onward and as the only supported pm since FreeBSD 10.0. It replaced what was known as the _pkg_*_ tools. The latter (or rather a re-implementation thereof) is still used on OpenBSD. On modern Solaris / illumos the "Image Packaging System" is used and the package manager for that is also called _pkg_.

Depending on how far Comtrya wants to take portability, it makes sense to consider this when choosing a provider name. Since Solaris and BSD use different manpage sections, a common differentiator is to use these when referring to either BSD's pkg(8) or Solaris' pkg(1). I'd probably call the provider "bsdpkg" or something; it's just helpful to be aware of the potential name clash upfront.

To allow for more rapid development, pkg(8) is not part of FreeBSD's base system. The system ships with a script, though, which transparently does the actual bootstrap and acts as a wrapper afterwards. FreeBSD's default shell is **not** bash but the C-shell variant called _tcsh_. This requires setting environment variables with env `$SOMETHING=$VALUE $COMMAND-TO-RUN` instead of `$SOMETHING=$VALUE $COMMAND` as you'd do with a bourne shell.

One thing to take into account regarding pkg(8) is that you are supposed to read the summary of actions to be taken before acknowledging it. The reason is that pkg deals with conflicts by proposing package removal of already installed packages. This is perfectly expected if e.g. installing Salt 300x on a system that still has Salt 2019 installed: As both install files to the same location, they are in conflict.

However in certain situations (due to interesting dependency issues) you can install a new package and end up having e.g. Libre Office or Firefox deleted. This definitely does not happen very often with the official FreeBSD repo, but somewhat more frequently for people rolling their own packages (which is a common thing to do). Since an automation tool cannot decide if pkg's proposal is the right thing for the system, this is a potential problem (I've had Salt screw up systems like this a couple of times over the years). Might make sense to choose safety by default here and allow for a special variable like "tolerate_remove" or something which defaults to "no".

To give the reader an impression of what that looks like, here's the output for trying to install the old version of Salt on a system that has a current version installed:
```
Checking integrity... done (1 conflicting)
  - py37-salt-2019-2019.2.8 conflicts with py37-salt-3002.5 on /usr/local/etc/salt/master.sample
Checking integrity... done (0 conflicting)
Conflicts with the existing packages have been found.
One more solver iteration is needed to resolve them.
The following 4 package(s) will be affected (of 0 checked):

Installed packages to be REMOVED:
        py37-salt: 3002.5

New packages to be INSTALLED:
        gmp: 6.2.1
        py37-pycrypto: 2.6.1_3
        py37-salt-2019: 2019.2.8

Number of packages to be removed: 1
Number of packages to be installed: 3

The operation will free 1 MiB.

Proceed with this action? [y/N]:
```

Here's some thoughts on an actual pkg provider:

Bootstrapping function:
- Check whether /usr/local/sbin/pkg exists
- Yes: Already bootstrapped, nothing to do
- No: Execute "env ASSUME_ALWAYS_YES=1 /usr/sbin/pkg bootstrap" to bootstrap without having to press "y"

Updating repository catalogue function (probably not even needed - pkg does this automatically if the catalogue is too old):
- Execute "pkg update"

Installing a package:
- If tolerating package removes on install: Execute "pkg install -y $PACKAGENAME"
- If not tolerating removes: Execute dry run "pkg install -n $PACKAGENAME"
- Parse output for "REMOVED", fail if encountered
- Execute "pkg install -y $PACKAGENAME"

Removing a package:
- Execute "pkg delete -y $PACKAGENAME"

**Comment by rawkode** [comment]
@kraileth Initial proof of concept is available here:

https://github.com/comtrya/comtrya/pull/72

For this, it always runs `pkg install -n package1 package2`. I will work on adding support for removal detection, but I don't know what that looks like yet.

I've got access to bare metal through my employer, so I plan to get a FreeBSD box up and running for further testing.

As `os_info` doesn't support BSD at this time, you'll need to explictly specify the provider, like this:

```yaml
actions:
- action: package.install
  name: curl
  provider: bsdpkg
```

**Comment by kraileth** [comment]
Awesome! Just gave it a try, but while building Comtrya I ran into an issue. I cloned the repository and checked out _feat/bsd-package_, then did `cargo build`.

Had to install the _perl5_ and _gmake_ packages so that _rust-openssl_ would build (I assume switching to vendored OpenSSL caused the need for this).

All dependencies build now, but for Comtrya itself I get 3 _E0053_ like this:

```
error[E0053]: method `bootstrap` has an incompatible type for trait     
  --> src/actions/package/providers/bsdpkg.rs:33:28
   |
33 |     fn bootstrap(&self) -> Result<(), crate::actions::ActionError> {
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `anyhow::Error`, found struct `ActionError`
   | 
  ::: src/actions/package/providers/mod.rs:82:28                                                                                                                              
   |                                                                                   
82 |     fn bootstrap(&self) -> Result<()>;
   |                            ---------- type in trait                
   |                                                                                   
   = note: expected fn pointer `fn(&BsdPkg) -> std::result::Result<_, anyhow::Error>`
              found fn pointer `fn(&BsdPkg) -> std::result::Result<_, ActionError>`
```

as well as 2 _E0277_ like this:

```
error[E0277]: the trait bound `anyhow::Error: StdError` is not satisfied
  --> src/actions/package/providers/bsdpkg.rs:45:11
   |
45 |         })?;
   |           ^ the trait `StdError` is not implemented for `anyhow::Error`
   |
   = note: required because of the requirements on the impl of `From<anyhow::Error>` for `ActionError`
   = note: required by `std::convert::From::from`
```

Tried to update Rust 1.48 to 1.50, but that did not help. Is there anything else that I'm missing?

**Comment by rawkode** [comment]
@kraileth Ooops. I think my build cache missed that somehow. Fixed!

**Comment by kraileth** [comment]
@rawkode It builds now and the provider seems to try to do its job. On a system deliberately missing the package manager I ran a simple manifest to install the bash shell. It outputs this (which looks good):

```
 WARN manifest_run{manifest="install"}:package.install{provider="BsdPkg"}: /usr/local/sbin/pkg is not available
 ERROR manifest_run{manifest="install"}: Provider unavailable
 ERROR manifest_run{manifest="install"}: Failed
```

Having pkg available, I re-ran Comtrya and got this:

```
 WARN manifest_run{manifest="install"}:package.install{provider="BsdPkg"}: Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
All repositories are up to date.
Updating database digests format: . done
The following 1 package(s) will be affected (of 0 checked):

New packages to be INSTALLED:
        bash: 5.1.4_1

Number of packages to be installed: 1

The process will require 8 MiB more space.
1 MiB to be downloaded.

 ERROR manifest_run{manifest="install"}:package.install{provider="BsdPkg"}: 
 ERROR manifest_run{manifest="install"}: Exit code: 1. Failed to run pkg install -n bash
 ERROR manifest_run{manifest="install"}: Failed
```

This also looks good. The problem here is that for whatever reason pkg(8) returns 1 after any dry-run! This is very weird and unexpected behavior. Will ask the FreeBSD developers about this. IMO it should return 0 if the package would have been installed if it hadn't been a dry-run and 1 only if something failed (e.g. there is no such package).

Could you maybe ignore the return value for the dry run - or treat 1 as acceptable in this very specific case? Then we'd have a temporary workaround to see if the actual functionality works as intended.

One more question: What would an action look like to trigger the bootstrap? I tried _package.bootstrap_ (and defined _provider: bsdpkg_) but this is obviously not correct. Is bootstrapping meant to be handled internally in bsdpkg?

**Comment by rawkode** [comment]
It should have bootstrapped 😭

That failed message is not very useful. I'll fix that. I'll also spin up a FreeBSD machine to do some testing myself, so I can have a bit more confidence in what I hand over to you

**Comment by kraileth** [comment]
@rawkode I see the problem. You call `/usr/local/sbin/pkg` - but that's unavailable until after the bootstrap. The wrapper script that does the bootstrap or passes stuff to the package manager, is `/usr/sbin/pkg`.

Quickly made the change, rebuilt and got a new error:

```
  WARN manifest_run{manifest="install"}:package.install{provider="BsdPkg"}: /usr/local/sbin/pkg is not available
  WARN manifest_run{manifest="install"}:package.install{provider="BsdPkg"}:bootstrap: 
 ERROR manifest_run{manifest="install"}:package.install{provider="BsdPkg"}:bootstrap: The package management tool is not yet installed on your system.
Please set ASSUME_ALWAYS_YES=yes environment variable to be able to bootstrap in non-interactive (stdin not being a tty)
```

**Comment by kraileth** [comment]
Ah, and the other problem is that it's "ASSUME_ALWAYS_YES" and not "ALWAYS_ASSUME_YES". I've tripped over this once, myself. The author of pkg is a Frenchman and thus English not his native language.

With this additional change bootstrapping works! :partying_face:

**Comment by rawkode** [comment]
Closed by #72
